### ISSUE TITLE
RESTEasy reactive / Undertow routing race

### ISSUE THREAD

**Issue by neusta4ail** [quarkusio/quarkus#14997]
**Bug Description**

While testing quarkus-resteasy-reactive (for REST) in combination with undertow (for websockets), I experienced four closely related problems:

*1. Race condition*: When activating quarkus-undertow[-websockets] + quarkus-resteasy-reactive[-*] extensions at the same time (i.e. to combine websockets with reactive JAX-RS), both will register routes/handlers at "/" by default in the Router. Undertow takes the lead due to its route's higher priority. If Undertow's servlet handler doesn't find a valid resource (which of course it won't), Undertow will fail the request with a 404 - NOT FOUND response without letting RESTEasy handle the request at all.

*2. Misleading error message*: If also enabling and requesting HTTP/2 then the error and response code will be misleading in this case (Status 5xx - "IllegalStateException: Request has already been read" rather than 404 - NOT FOUND).

*3. Performance/Configuration*: Setting quarkus.http.root-path to a non-default value leaves the RESTEasy handler's top-level route at "/" and handles the subroute somewhere inside the handler. IMHO it would be better to have the root path in the top-level route which would enable the router to discover much earlier which paths should (not) be handled by resteasy (see the equivalent behavior of quarkus.servlet.context-path). This would also improve isolation of handlers and might make it easier to avoid the above race condition in code.

*4. Documentation*: The resteasy-docs contain a reference to quarkus.resteasy.path  which didn't do anything in my case and doesn't seem to be referenced any more (neither in reasteasy nor in resteasy-reactive source code). quarkus.http.root-path solves the problem to some extend but a resteasy specific path (see equivalent quarkus.servlet.context-path) would be better IMHO as quarkus.http.root-path has further side effects which may not be intended if one simply wants to set a top-level route for the resteasy handler. It would also be great if the effects of quarkus.http.root-path on RESTEasy and the non-effect on Undertow would be better documented.

**Expected behavior**

Possible ideas (not sure whether they make sense...):
* In case of a routing conflict between the different handlers, a good error message could inform the user that quarkus.servlet.context-path should be set to a non-default value to let resteasy-reactive and undertow co-operate. This will also improve routing performance and handler isolation as a much shorter code path will be taken to find out which (single) handler should handle the request.
* The servlet-part of undertow could be disabled (with an info-level log message) if resteasy-reactive is configured. Most users will probably not use the servlet part of undertow in this case.
* The undertow and resteasy extensions could co-operate by not failing the request unless both had a chance to try to resolve it (slower and more complicated possible interactions that need to be managed - but would allow the two extensions to handle the same top-level URL namespace). IMO opinion, not separating the namespaces doesn't make much sense, though.
* RESTEasy's route/handler could be inserted with a higher priority than undertow's - but then the same race condition and co-operation problems exist the other way round.

**Actual behavior**

Requests to valid routes fail with a misleading error message (404 - NOT FOUND in case of HTTP/1.1 or IllegalStateException in case of HTTP/2) - see above.

**To Reproduce**

Activate both undertow and resteasy-reactive and try to access an arbitrary JAX-RS endpoint that should be handled by resteasy.

**Configuration**
```properties
# Default config will lead to the above race condition.
# The servlet (aka Undertow) is only used to serve WebSocket traffic.
# It needs to be placed on a separate route to avoid conflict with RESTEasy Reactive.
quarkus.servlet.context-path="/" # vs. quarkus.servlet.context-path=/socket - see bug description
quarkus.http.http2=enabled # vs. disabled - see bug description
```

**Screenshots**
 - n/a
 
**Environment (please complete the following information):**
 - all environments

**Additional context**
 - n/a

**Comment by geoand** [comment]
@stuartwdouglas @FroMage I wonder if we should just disallow RESTEasy Reactive + Undertow for now as they don't seem to play well at alll together...

**Comment by fgrandel-neusta** [comment]
@geoand - That would be a problem for us as long as there is no Websocket solution within resteasy-reactive.

**Comment by geoand** [comment]
> @geoand - That would be a problem for us as long as there is no Websocket solution within resteasy-reactive and there is a simple configurative workaround that solves the problem (see the bug description).

I am pretty sure that @stuartwdouglas has something in the works

**Comment by stuartwdouglas** [comment]
I have something in the works, I had put it on hold due to not wanting to conflict with the Vert.x 4 work but it might be worth just dealing with it, and if it causes problems I will just deal with them: https://github.com/quarkusio/quarkus-http/pull/55.

For now you can add quarkus-resteasy-reactive-servlet to your project and run on Undertow. It won't be quite as performant but still should be faster than classic RESTEasy.

**Comment by fgrandel-neusta** [comment]
Hi @stuartwdouglas, thanks for your support and feedback.

We do have a working configurative workaround, though, simply by setting quarkus.servlet.context-path to some non-default root path other than "/". This will change Undertow recorder's top-level route to a different value and will thereby let quarkus-resteasy-reactive-jackson (in our case) deal with the requests on other routes at full performance.

I guess that this workaround will be fine for most users until someone has the time to look into this. So from our side no need to rush anything - just wanted to let you know that the current default configuration, documentation and error messages will lead people into the wrong direction unless they look at the source code.

Maybe a simple adjustment of the docs + some better error messages will do the trick for the moment being?

**Comment by gsmet** [comment]
The new Vert.x based websockets support will be available in 1.13 so I think we can close this one.
